src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -364,6 +365,51 @@ FileWatcherCertificateProvider::ReadIdentityKeyCertPairFromFiles(
   return absl::nullopt;
 }
 
+absl::Status PrivateKeyPublicKeyMatches(const std::string& private_key_string,
+                                        const std::string& cert_string) {
+  if (cert_string.empty()) {
+    return absl::InvalidArgumentError("Certificate string is empty!");
+  }
+  // -1 indicates a null-terminated string
+  BIO* private_string_bio(BIO_new_mem_buf(cert_string.c_str(), -1));
+  if (private_string_bio == nullptr) {
+    return absl::InvalidArgumentError(
+        "Conversion from certificate string to BIO failed!");
+  }
+  X509* x509(PEM_read_bio_X509(private_string_bio, nullptr, nullptr, nullptr));
+  BIO_free(private_string_bio);
+  if (x509 == nullptr) {
+    return absl::InvalidArgumentError(
+        "Conversion from PEM string to X509 failed!");
+  }
+  EVP_PKEY* public_key(X509_get_pubkey(x509));
+  if (public_key == nullptr) {
+    return absl::InvalidArgumentError(
+        "Extraction of public key from x.509 certificate failed!");
+  }
+  if (private_key_string.empty()) {
```

Comment:
Move this to start of the code. So you exit early if the value is unavailable.

Reply:
Noted!